From 4b5cda5a2c6ca613db5129d50ffffce2604ad9eb Mon Sep 17 00:00:00 2001
From: Yash Datta <Yash.Datta@guavus.com>
Date: Mon, 1 Jun 2015 14:21:53 -0700
Subject: [PATCH] PARQUET-151: Skip writing _metadata file in case of no
 footers since schema cannot be determined.

This fixes npe seen during mergeFooters in such a case.
 For this scenario onus of writing any summary files lies with the caller (It might have some global schema available) So for example spark does it when persisting empty RDD.

Author: Yash Datta <Yash.Datta@guavus.com>

Closes #205 from saucam/footer_bug and squashes the following commits:

b2b3ddf [Yash Datta] PARQUET-151: Skip writing _metadata file in case of no footers since schema cannot be determined. This fixes npe seen during mergeFooters in such a case.              For this scenario onus of writing any summary files lies with the caller (It might have some global schema available)
---
 .../org/apache/parquet/hadoop/ParquetOutputCommitter.java    | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetOutputCommitter.java b/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetOutputCommitter.java
index a1589c0de0..9a0930a904 100644
--- a/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetOutputCommitter.java
+++ b/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetOutputCommitter.java
@@ -54,6 +54,11 @@ public static void writeMetaDataFile(Configuration configuration, Path outputPat
         final FileSystem fileSystem = outputPath.getFileSystem(configuration);
         FileStatus outputStatus = fileSystem.getFileStatus(outputPath);
         List<Footer> footers = ParquetFileReader.readAllFootersInParallel(configuration, outputStatus);
+        // If there are no footers, _metadata file cannot be written since there is no way to determine schema!
+        // Onus of writing any summary files lies with the caller in this case.
+        if (footers.isEmpty()) {
+          return;
+        }
         try {
           ParquetFileWriter.writeMetadataFile(configuration, outputPath, footers);
         } catch (Exception e) {
